File: AlgorithmLcr/Try2-Known component types, quantities and parameters/netgraph_id/filters.py
"""Pruning funnel (DESIGN.md section 7).

With known component values every candidate is *exactly* evaluable at any
frequency, so the cheapest complete filter is a small set of probe
frequencies (band edges + middle): a wrong wiring of exact-valued
components is wrong by O(1) relative error somewhere in the band, while
the true wiring sits at the noise floor (~sigma_rel).  The funnel keeps
every candidate whose probe RSS is within `funnel_ratio` of the running
best (generous by construction: it can only drop candidates that are wrong
by more than ~sqrt(funnel_ratio) relative, far beyond any noise level),
then fully evaluates the survivors.  This subsumes the classical
DC/HF-asymptote checks (Z(0), Z(inf)) which are additionally reported for
transparency.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from .components import ComponentSet
from .enumerate import enumerate_structures, iter_assignments
from .graph import Network, Structure
from .nodal import StructureStamps
from .metric import weighted_rss


def coarse_indices(M: int, n_points: int = 3) -> list[int]:
    """Probe frequencies: band edges + interior, spread over the band."""
    if M <= n_points:
        return list(range(M))
    return sorted(set([0, M - 1] + [round(i * (M - 1) / (n_points - 1)) for i in range(n_points)]))


@dataclass
class FunnelState:
    """Streaming funnel accumulator across structures."""

    compset: ComponentSet
    s: np.ndarray                    # full angular frequencies (M,)
    z: np.ndarray                    # measurements (M,)
    w: np.ndarray                    # weights (M,)
    probe_idx: list[int]
    funnel_ratio: float
    min_keep: int
    batch_size: int = 4096

    best_probe: float = np.inf
    kept: list[tuple] = field(default_factory=list)   # (probe_rss, Network)
    n_total: int = 0
    n_probe_evaluated: int = 0

    def _update(self, nets: list[Network], probe_rss: np.ndarray) -> None:
        for net, prss in zip(nets, probe_rss):
            self.n_total += 1
            if prss < self.best_probe:
                self.best_probe = float(prss)
        self.n_probe_evaluated += len(nets)
        thr = self.best_probe * self.funnel_ratio
        for net, prss in zip(nets, probe_rss):
            if prss <= thr:
                self.kept.append((float(prss), net))

    def final_keep(self) -> list[Network]:
        """Survivors after the global best is known; always keep some."""
        thr = self.best_probe * self.funnel_ratio
        kept = [net for prss, net in self.kept if prss <= thr]
        if len(kept) < self.min_keep:
            self.kept.sort(key=lambda t: t[0])
            kept = [net for _, net in self.kept[:self.min_keep]]
        return kept


def run_funnel(compset: ComponentSet, s: np.ndarray, z: np.ndarray,
               w: np.ndarray, config) -> FunnelState:
    """Enumerate all structures, stream assignments, probe-filter them."""
    E = compset.n
    structures = enumerate_structures(E, allow_dead=config.allow_dead)
    probe_idx = coarse_indices(len(s), config.coarse_points)
    state = FunnelState(compset=compset, s=s, z=z, w=w, probe_idx=probe_idx,
                        funnel_ratio=config.funnel_ratio,
                        min_keep=config.funnel_min_keep,
                        batch_size=config.batch_size)
    s_probe = s[probe_idx]
    z_probe = z[probe_idx]
    w_probe = w[probe_idx]

    for structure in structures:
        stamps = StructureStamps.build(structure, compset)
        batch: list[tuple[int, ...]] = []
        for assign in iter_assignments(structure, compset):
            batch.append(assign)
            if len(batch) >= state.batch_size:
                _probe_batch(state, stamps, batch, s_probe, z_probe, w_probe)
                batch = []
        if batch:
            _probe_batch(state, stamps, batch, s_probe, z_probe, w_probe)
    return state


def _probe_batch(state: FunnelState, stamps: StructureStamps,
                 batch: list[tuple[int, ...]], s_probe, z_probe,
                 w_probe) -> None:
    assigns = np.array(batch, dtype=np.intp)
    z_model = stamps.z_full(assigns, s_probe)         # (N, P)
    probe_rss = weighted_rss(z_probe, z_model, w_probe)
    nets = [Network(structure=stamps.structure, assign=tuple(a)) for a in batch]
    state._update(nets, probe_rss)

File: AlgorithmLcr/Try2-Known component types, quantities and parameters/cppversion/src/filters.cpp
#include "filters.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <set>

namespace ng {

long roundHalfEven(double x) {
    double f = std::floor(x);
    double d = x - f;
    if (d > 0.5) return (long)f + 1;
    if (d < 0.5) return (long)f;
    return (((long)f) % 2 == 0) ? (long)f : (long)f + 1;
}

std::vector<int> coarseIndices(int M, int nPoints) {
    if (M <= nPoints) {
        std::vector<int> out(M);
        for (int i = 0; i < M; ++i) out[i] = i;
        return out;
    }
    std::set<int> uq;
    uq.insert(0);
    uq.insert(M - 1);
    for (int i = 0; i < nPoints; ++i)
        uq.insert((int)roundHalfEven((double)i * (M - 1) / (nPoints - 1)));
    return std::vector<int>(uq.begin(), uq.end());
}

void FunnelState::update(const std::vector<Network>& nets,
                         const std::vector<double>& probeRss) {
    for (size_t i = 0; i < nets.size(); ++i) {
        nTotal += 1;
        if (probeRss[i] < bestProbe) bestProbe = probeRss[i];
    }
    nProbeEvaluated += (long)nets.size();
    double thr = bestProbe * funnelRatio;
    for (size_t i = 0; i < nets.size(); ++i)
        if (probeRss[i] <= thr) kept.push_back({probeRss[i], nets[i]});
}

std::vector<Network> FunnelState::finalKeep() const {
    double thr = bestProbe * funnelRatio;
    std::vector<Network> out;
    for (const auto& [prss, net] : kept)
        if (prss <= thr) out.push_back(net);
    if ((int)out.size() < minKeep) {
        std::vector<std::pair<double, Network>> sortedKept = kept;
        std::stable_sort(sortedKept.begin(), sortedKept.end(),
                         [](const auto& a, const auto& b) { return a.first < b.first; });
        out.clear();
        for (int i = 0; i < minKeep && i < (int)sortedKept.size(); ++i)
            out.push_back(sortedKept[i].second);
    }
    return out;
}

FunnelState runFunnel(const ComponentSet& compset, const std::vector<Complex>& s,
                      const std::vector<Complex>& z, const std::vector<double>& w,
                      int coarsePoints, double funnelRatio, int minKeep, int batchSize,
                      bool allowDead) {
    const int E = compset.n();
    const auto& structures = enumerateStructures(E, allowDead);
    std::vector<int> probeIdx = coarseIndices((int)s.size(), coarsePoints);
    FunnelState state{compset, s, z, w, std::move(probeIdx), funnelRatio, minKeep,
                      batchSize, std::numeric_limits<double>::infinity(), {}, 0, 0};
    std::vector<Complex> sProbe, zProbe;
    std::vector<double> wProbe;
    for (int idx : state.probeIdx) {
        sProbe.push_back(s[idx]);
        zProbe.push_back(z[idx]);
        wProbe.push_back(w[idx]);
    }

    for (const auto& structure : structures) {
        StructureStamps stamps = StructureStamps::build(structure, compset);
        std::vector<std::vector<int>> batch;
        batch.reserve((size_t)state.batchSize);
        iterAssignments(structure, compset,
                        [&](const std::vector<int>& assign) {
                            batch.push_back(assign);
                            if ((int)batch.size() >= state.batchSize) {
                                std::vector<std::vector<Complex>> zm =
                                    stamps.zFull(batch, sProbe);
                                std::vector<double> prss =
                                    weightedRssBatch(zProbe, zm, wProbe);
                                std::vector<Network> nets;
                                nets.reserve(batch.size());
                                for (const auto& a : batch)
                                    nets.push_back(Network{structure, a});
                                state.update(nets, prss);
                                batch.clear();
                            }
                        });
        if (!batch.empty()) {
            std::vector<std::vector<Complex>> zm = stamps.zFull(batch, sProbe);
            std::vector<double> prss = weightedRssBatch(zProbe, zm, wProbe);
            std::vector<Network> nets;
            nets.reserve(batch.size());
            for (const auto& a : batch) nets.push_back(Network{structure, a});
            state.update(nets, prss);
        }
    }
    return state;
}

}  // namespace ng
